server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -1433,11 +1445,59 @@ private static RemoteTransportException wrapRemoteClusterFailure(String clusterA
                     new String[] { shardId.getIndexName() },
                     originalIndices.indicesOptions()
                 );
-                iterators.add(
-                    new SearchShardIterator(localClusterAlias, shardId, targetNodes, finalIndices, perNode.getSearchContextId(), keepAlive)
-                );
+                boolean passesIndexCheck = shouldFilterIndices == false || passesIndexCheck(shardId.getIndexName(), indices);
+                boolean passesShardCheck = shardIds.isEmpty() || shardIds.contains(shardId.id());
+                if (passesIndexCheck && passesShardCheck) {
+                    iterators.add(
+                        new SearchShardIterator(
+                            localClusterAlias,
+                            shardId,
+                            targetNodes,
+                            finalIndices,
+                            perNode.getSearchContextId(),
+                            keepAlive
+                        )
+                    );
+                } else {
+                    throw new IllegalArgumentException(
+                        "Got indices: " + indices + " and shards: " + shardIds + " but actual was " + shardId.getIndex()
+                    );
+                }
             }
         }
         return iterators;
     }
+
+    private static boolean passesIndexCheck(String indexNameFromPit, Set<String> indexNamesFromRequest) {
+        if (indexNamesFromRequest.contains(indexNameFromPit)) {
+            return true;
+        }
+        for (String indexNameFromRequest : indexNamesFromRequest) {
+            if (indexNameFromPit.startsWith(".ds-" + indexNameFromRequest)) {
```

Comment:
We don't want to merge this. This is just to get everything working with data streams.